### PR TITLE
Add a description on how a leader replica executes a command

### PR DESCRIPTION
**Pull request by kkaneda**
I thought that this could be useful for someone to understand how `Replica` processes commands.

### CODE REVIEW

File: docs/design.md [hunk 1]
```diff
@@ -860,6 +860,50 @@ already leading), though some care should be taken that Range leadership is
...
+key(s) belongs to. When receiving a command, a RoachNode looks up a
+range by the specified Range ID and checks if the range is still
+responsible for the keys. When all the keys do not belong to the
+range, the RoachNode returns an error so that the client will retry
+and send a request to a correct range.
+
+When all the keys belong to the range, the RoachNode attempts to
+process the command. If the command is an inconsistent read-only
+command, it is processed immediately. If the command is a consistent
+read/write, the command is executed when both of the following
```
Comment (bdarnell) [inline docs/design.md]: s#read/write#read or a write#

File: docs/design.md [hunk 2]
```diff
@@ -860,6 +860,50 @@ already leading), though some care should be taken that Range leadership is
...
+- There are no other running commands whose keys overlap with
+the submitted command and cause read/write conflict.
+
+When the first condition does not met, the replica attempts to acquire
+a lease or returns an error so that the client will redirect the
+command to the current leader. The second condition guarantees that
+consistent read/write commands for a given key are sequentially
+executed.
+
+When the above two conditions is met, the leader replica commits the
```
Comment (bdarnell) [inline docs/design.md]: Only writes go to the raft log. Consistent reads are processed on the leader as soon as there are no other conflicting commands.

File: docs/design.md [hunk 3]
```diff
@@ -860,6 +860,50 @@ already leading), though some care should be taken that Range leadership is
...
+command into the Raft log and updates its followers (and every replica
+will execute the same command). All the commands except leader lease
+request produce deterministic results so that the range replicas keep
```
Comment (bdarnell) [inline docs/design.md]: The leader lease command is also deterministic (as it must be, for all replicas to agree on who has the lease).

File: docs/design.md (hunk 1 above)
Comment (kkaneda) [inline docs/design.md]: Fixed.

File: docs/design.md (hunk 2 above)
Comment (kkaneda) [inline docs/design.md]: Fixed.

File: docs/design.md (hunk 3 above)
Comment (kkaneda) [inline docs/design.md]: Ah, yes (except to the update to the low water mark of the timestamp cache?). Fixed.

**Comment by kkaneda** [comment]
Thanks for the review, @bdarnell ! Addressed the comments.

File: docs/design.md [hunk 4]
```diff
@@ -860,6 +860,50 @@ already leading), though some care should be taken that Range leadership is
...
+process the command. If the command is an inconsistent read-only
+command, it is processed immediately. If the command is a consistent
+read or a write, the command is executed when both of the following
+conditions hold:
+
+- The range replica has a leader lease.
+- There are no other running commands whose keys overlap with
+the submitted command and cause read/write conflict.
+
+When the first condition does not met, the replica attempts to acquire
```
Comment (mrtracy) [inline docs/design.md]: s/does not/is not/

File: docs/design.md [hunk 5]
```diff
@@ -860,6 +860,50 @@ already leading), though some care should be taken that Range leadership is
...
 transitions.
 
+## Command Execution Flow
+
+This subsection describes how a leader replica processes a read/write
+command in more details. Each command specifies (1) a key (or a range
+of keys) that the command accesses and (2) the ID of a range which the
+key(s) belongs to. When receiving a command, a RoachNode looks up a
+range by the specified Range ID and checks if the range is still
+responsible for the keys. When all the keys do not belong to the
```
Comment (mrtracy) [inline docs/design.md]: s/for the keys/for the *supplied* keys/

File: docs/design.md [hunk 6]
```diff
@@ -860,6 +860,50 @@ already leading), though some care should be taken that Range leadership is
...
+- There are no other running commands whose keys overlap with
+the submitted command and cause read/write conflict.
+
+When the first condition does not met, the replica attempts to acquire
+a lease or returns an error so that the client will redirect the
+command to the current leader. The second condition guarantees that
+consistent read/write commands for a given key are sequentially
+executed.
+
+When the above two conditions is met, the leader replica processes the
```
Comment (mrtracy) [inline docs/design.md]: s/is met/are met/

File: docs/design.md [hunk 7]
```diff
@@ -860,6 +860,50 @@ already leading), though some care should be taken that Range leadership is
...
 
+## Command Execution Flow
+
+This subsection describes how a leader replica processes a read/write
+command in more details. Each command specifies (1) a key (or a range
+of keys) that the command accesses and (2) the ID of a range which the
+key(s) belongs to. When receiving a command, a RoachNode looks up a
+range by the specified Range ID and checks if the range is still
+responsible for the keys. When all the keys do not belong to the
+range, the RoachNode returns an error so that the client will retry
```
Comment (mrtracy) [inline docs/design.md]: s/When all the keys do not belong/If any of the keys do not belong/

**Comment by kkaneda** [comment]
Thanks, @mrtracy ! Fixed all.

File: docs/design.md (hunk 5 above)
Comment (kkaneda) [inline docs/design.md]: Fixed.

File: docs/design.md (hunk 7 above)
Comment (kkaneda) [inline docs/design.md]: Fixed.

File: docs/design.md (hunk 4 above)
Comment (kkaneda) [inline docs/design.md]: Fixed.

File: docs/design.md (hunk 6 above)
Comment (kkaneda) [inline docs/design.md]: Fixed.

**Comment by bdarnell** [comment]
LGTM

**Comment by mrtracy** [comment]
LGTM

**Pull request merged**